### ISSUE TITLE
I have an issue deploying a tensorflow contained exe

### ISSUE THREAD

**Issue by arcgos** [mherrmann/fbs-tutorial#12]
Fbs is working properly, and python -m fbs run is also working for my main script, but when I freeze the model , there are many warnings and when I run the exe file in target then it shows failed to execute script main 

I am having the following error
_________________________________________
87082 WARNING: lib not found: lib_blas_su.2UNBVF3XUJ6LXUE74AWX6SW54UVZEM5P.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\linalg\cython_lapack.cp36-win_amd64.pyd
87146 WARNING: lib not found: lib_blas_su.2UNBVF3XUJ6LXUE74AWX6SW54UVZEM5P.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\linalg\cython_blas.cp36-win_amd64.pyd
87257 WARNING: lib not found: libnnls.IXEEHJUCGHJL42YZEM6UIEMROJWXHMLJ.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\optimize\_nnls.cp36-win_amd64.pyd
87306 WARNING: lib not found: libslsqp_op.NNY57ZXZ43A4RH3YWFA7BKHP5PC2K3I5.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\optimize\_slsqp.cp36-win_amd64.pyd
87370 WARNING: lib not found: libchkder.6HLXPVTQJEGRZGLI5DFRMNW3SS76BHP6.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\optimize\_minpack.cp36-win_amd64.pyd
87434 WARNING: lib not found: libcobyla2.JEGTSUUFJ7DFXWZN5PAYZTTLBDATC4WD.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\optimize\_cobyla.cp36-win_amd64.pyd
87498 WARNING: lib not found: liblbfgsb.LTROWG5MXPUSP7N2YCW3QUHJE37DB7K5.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\optimize\_lbfgsb.cp36-win_amd64.pyd
87549 WARNING: lib not found: libdcsrch.I2AOPDCXAPDRFNPWY55H5UE7XZSU5CVN.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\optimize\minpack2.cp36-win_amd64.pyd
87634 WARNING: lib not found: lib_arpack-.ZSFN7AZ4NRWSAUZTEZLC7L2IVIRDDNFR.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\sparse\linalg\eigen\arpack\_arpack.cp36-win_amd64.pyd
87734 WARNING: lib not found: libgetbreak.WYF32CNQCVMHFKRVSHNE3EGBPQL7SWIB.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\sparse\linalg\isolve\_iterative.cp36-win_amd64.pyd
88283 WARNING: lib not found: libdfitpack.KTCF3EOE66VRDKN45KBQA4VBAIS552IF.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\interpolate\dfitpack.cp36-win_amd64.pyd
88336 WARNING: lib not found: libbispeu.5N2XSD7URZS4WTOSLTOG4DDMA4HGB46U.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\interpolate\_fitpack.cp36-win_amd64.pyd
88384 WARNING: lib not found: libblkdta00.EBMXOKVWTYK6NSUJLSZFNQ6CEWS6SU3V.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\integrate\lsoda.cp36-win_amd64.pyd
88432 WARNING: lib not found: libdop853.6TJTQZW3I3Q3QIDQHEOBEZKJ3NYRXI4B.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\integrate\_dop.cp36-win_amd64.pyd
88480 WARNING: lib not found: libvode.M2PR7ET5DWOC2TSRZM44MAX2MG4YDRX3.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\integrate\vode.cp36-win_amd64.pyd
88528 WARNING: lib not found: libdqag.JBSUT27DU6KSWWLH5EAFVANLG5T4CPYH.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\integrate\_quadpack.cp36-win_amd64.pyd
88592 WARNING: lib not found: libblkdta00.EBMXOKVWTYK6NSUJLSZFNQ6CEWS6SU3V.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\integrate\_odepack.cp36-win_amd64.pyd
88653 WARNING: lib not found: libspecfun.BHLTWMBI4EYWDACZN4DQUESSDJRJNGEL.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\special\specfun.cp36-win_amd64.pyd
88704 WARNING: lib not found: libdgamln.D35ZEOCATK4OZQSX5UEWEF62EM2ECCPX.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\special\_ufuncs.cp36-win_amd64.pyd
88772 WARNING: lib not found: libdet.YKP4TFVQQVNWGO5ZEH7E6CRNLZ42FK5D.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\linalg\_flinalg.cp36-win_amd64.pyd
88836 WARNING: lib not found: libwrap_dum.CLCNYV3KUDVQF2VOOGW7JOOKWHJ2FDXT.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\linalg\_flapack.cp36-win_amd64.pyd
88899 WARNING: lib not found: libwrap_dum.ASTX55GIE5U6BF4YKSWUU6TBMWRPJ2I4.gfortran-win_amd64.dll dependency of e:\fbs-tutorial-master\fbs-tutorial-master\venv\lib\site-packages\scipy\linalg\_fblas.cp36-win_amd64.pyd

**Comment by mherrmann** [comment]
Does it work when you copy the .dlls mentioned in the error next to the .exe produced by `fbs freeze`?

**Comment by arcgos** [comment]
I have put most of those dlls , yet same issue failed to execute script main
For precaution I have put all dlls and pyd files present in scipy

**Comment by mherrmann** [comment]
And do you have the dlls somewhere else maybe? In the worst case, do a full system search(?)

**Comment by ahgamut** [comment]
I have a similar issue with using scipy.

Adding `location_of_my_python_env\Lib\site-packages\scipy\extra-dll\` to the `PATH` environment variable, and then executing `python -m fbs freeze`, suppresses the warning(s), but the executable still crashes with `Failure to execute script main`.

**Comment by mherrmann** [comment]
Can you create a minimum example that reproduces the issue? I'm afraid I don't have experience with scipy or tensorflow.

**Comment by ahgamut** [comment]
Given below is a minimal example that throws up the error on my computer:

1. I followed the tutorial: downloaded the repo, set up a virtual env (via cmd, not any IDE). I downloaded Python separately as well, I am not using Anaconda.
2. I added libraries to `requirements.txt`
	```
	fbs==0.1.3
	PyQt5==5.9.2
	PyInstaller==3.3.1
	numpy
	scipy
	matplotlib
	scikit-image
	```
3. Set up the virtual env with no issue, able to reproduce the example program with `success.jpg`
4. Then I modified `application_context_2.py` to produce a random number. No changes to any other file.

5. `application_context_2.py` looks like this:

```python

from fbs_runtime.application_context import ApplicationContext, \
    cached_property
from PyQt5.QtGui import QPixmap
from PyQt5.QtWidgets import QApplication, QMainWindow, QLabel, QPushButton, QHBoxLayout, QVBoxLayout, QWidget

from numpy.random import rand
#from scipy.stats import dgamma

class MyMainWindow(QMainWindow):
    def __init__(self):
        QMainWindow.__init__(self, parent = None)
        self.setWindowTitle("This is my GUI")
        self.rand_button = QPushButton('Generate random value', parent=self)
        self.rand_button.clicked.connect(self.listener)
        
    def arrange_gui(self):
        self.setCentralWidget(self.rand_button)
    
    def listener(self):        
        r = rand(1) # importing from numpy has no issue
        
        # r = dgamma.rvs(1.1, size=1) 
        # If I uncomment the above line (and the import) it runs on `python -m fbs run`, but not after freezing
        
        self.setWindowTitle(str(r))

class AppContext(ApplicationContext):
    def run(self):
        self.main_window.show()
        return self.app.exec_()
    @cached_property
    def main_window(self):
        result = MyMainWindow()
        return result
```



FWIW, I've attached what I did on `cmd` as well:

```cmd

(mvenv) E:\fbs-tutorial-master>echo "using only numpy"
"using only numpy"

(mvenv) E:\fbs-tutorial-master>python -m fbs run

(mvenv) E:\fbs-tutorial-master>echo "run works"
"run works"

(mvenv) E:\fbs-tutorial-master>python -m fbs freeze
WARNING: The output directory "E:\fbs-tutorial-master\target\Tutorial" and ALL ITS CONTENTS will be REMOVED! Continue? (y/n)y

(mvenv) E:\fbs-tutorial-master>cd target\Tutorial

(mvenv) E:\fbs-tutorial-master\target\Tutorial>Tutorial.exe

(mvenv) E:\fbs-tutorial-master\target\Tutorial>echo "Works as expected"
"Works as expected"

(mvenv) E:\fbs-tutorial-master\target\Tutorial>cd ..\..

(mvenv) E:\fbs-tutorial-master>python -m fbs run

(mvenv) E:\fbs-tutorial-master>"scipy works on run"
'"scipy works on run"' is not recognized as an internal or external command,
operable program or batch file.

(mvenv) E:\fbs-tutorial-master>echo "scipy works on run"
"scipy works on run"

(mvenv) E:\fbs-tutorial-master>python -m fbs freeze
40966 WARNING: lib not found: libwrap_dum.Y2ZSI7BHNFII36DED275BJU6FIUD2ROF.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\linalg\
_fblas.cp35-win32.pyd
41097 WARNING: lib not found: libwrap_dum.5GD2LRZQ4JEZWXIKBNM3DBEABCEUS7E7.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\linalg\
_flapack.cp35-win32.pyd
41211 WARNING: lib not found: libdet.Q6KQSP2572HW5FK6O3L7QJ6NV2QC46AS.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\linalg\_flin
alg.cp35-win32.pyd
41354 WARNING: lib not found: libspecfun.I7OMDT5L33XVQM2MTW5AACQJUBYEOUFN.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\special\
specfun.cp35-win32.pyd
41470 WARNING: lib not found: libdgamln.RNMMLYNRRSOBQTZMLBVQS5336MUR2KBM.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\special\_
ufuncs.cp35-win32.pyd
41607 WARNING: lib not found: lib_arpack-.GCORIMVZFNUUKF2PPB5GBEYXA2DICUB2.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\sparse\
linalg\eigen\arpack\_arpack.cp35-win32.pyd
41739 WARNING: lib not found: libgetbreak.UXNRK2F7GAG3K6DO3WEUC7MSI4YQGOPD.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\sparse\
linalg\isolve\_iterative.cp35-win32.pyd
42581 WARNING: lib not found: libdfitpack.DAK6MWNYDQZNNBSTJB7HILANAHXEOEK5.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\interpo
late\dfitpack.cp35-win32.pyd
42701 WARNING: lib not found: libbispeu.OAANPWJKKXZRFOCA7BPAXPEXKORTJQMF.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\interpola
te\_fitpack.cp35-win32.pyd
42822 WARNING: lib not found: libblkdta00.EXU2PLJ5B7YHAI66VSZ4X25Z7GIWUJLZ.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\integra
te\lsoda.cp35-win32.pyd
42941 WARNING: lib not found: libdop853.ZDB77F5S63EPO7WWG3LICXZSBW2LFM2N.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\integrate
\_dop.cp35-win32.pyd
43065 WARNING: lib not found: libvode.ZGVCPKWVS3YQF2IPELFOMCKD4AMMHI65.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\integrate\v
ode.cp35-win32.pyd
43180 WARNING: lib not found: libdqag.7J2OZWWCUYCYBJNAYBJYYZO7VCVPCGKT.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\integrate\_
quadpack.cp35-win32.pyd
43301 WARNING: lib not found: libblkdta00.EXU2PLJ5B7YHAI66VSZ4X25Z7GIWUJLZ.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\integra
te\_odepack.cp35-win32.pyd
43425 WARNING: lib not found: libnnls.GX3LBH56JRQ7JMAOG7UBTDQJYFS6BPRN.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\optimize\_n
nls.cp35-win32.pyd
43544 WARNING: lib not found: libslsqp_op.PY4IBHJ4EFVQXCMTZF6MCCENNULFVBJB.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\optimiz
e\_slsqp.cp35-win32.pyd
43667 WARNING: lib not found: libchkder.ACKDBOJBSJX62PBVWDOLR7SBBWNGB4T2.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\optimize\
_minpack.cp35-win32.pyd
43803 WARNING: lib not found: libcobyla2.MZIYBZPTVTYIHRT7JCIZN4FNMM5IPWLD.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\optimize
\_cobyla.cp35-win32.pyd
43928 WARNING: lib not found: liblbfgsb.BTS3IJZAI5MR3XZTEFZYKGXI3QWL3ZLJ.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\optimize\
_lbfgsb.cp35-win32.pyd
44046 WARNING: lib not found: libdcsrch.CTD6WPTKUW4MYYPI2KYOJ4427ZGGSXEQ.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\optimize\
minpack2.cp35-win32.pyd
44173 WARNING: lib not found: lib_blas_su.I7T7R5XGQYU2YRID4LTK2GOFCSIKSOTB.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\linalg\
cython_lapack.cp35-win32.pyd
44309 WARNING: lib not found: lib_blas_su.I7T7R5XGQYU2YRID4LTK2GOFCSIKSOTB.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\linalg\
cython_blas.cp35-win32.pyd
44461 WARNING: lib not found: libmvndst.FOQEXFKZWBDPO7HSQQJGMDNRVPGTHK7M.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\stats\mvn
.cp35-win32.pyd
44588 WARNING: lib not found: libansari.Q4BAGRNANLWD2YZJOKYPOAUIOLXW2LXK.gfortran-win32.dll dependency of e:\fbs-tutorial-master\mvenv\lib\site-packages\scipy\stats\sta
tlib.cp35-win32.pyd
WARNING: The output directory "E:\fbs-tutorial-master\target\Tutorial" and ALL ITS CONTENTS will be REMOVED! Continue? (y/n)y

(mvenv) E:\fbs-tutorial-master>cd target\Tutorial

(mvenv) E:\fbs-tutorial-master\target\Tutorial>Tutorial.exe

(mvenv) E:\fbs-tutorial-master\target\Tutorial>echo "Failed to execute script main"
"Failed to execute script main"

(mvenv) E:\fbs-tutorial-master\target\Tutorial>set PATH=%PATH%;E:\fbs-tutorial-master\mvenv\Lib\site-packages\scipy\extra-dll\;

(mvenv) E:\fbs-tutorial-master\target\Tutorial>echo "added location of scipy dlls to PATH"
"added location of scipy dlls to PATH"

(mvenv) E:\fbs-tutorial-master\target\Tutorial>cd ..\..

(mvenv) E:\fbs-tutorial-master>python -m fbs freeze
WARNING: The output directory "E:\fbs-tutorial-master\target\Tutorial" and ALL ITS CONTENTS will be REMOVED! Continue? (y/n)y

(mvenv) E:\fbs-tutorial-master>echo "No dll warnings now"
"No dll warnings now"

(mvenv) E:\fbs-tutorial-master>cd target\Tutorial

(mvenv) E:\fbs-tutorial-master\target\Tutorial>Tutorial.exe

(mvenv) E:\fbs-tutorial-master\target\Tutorial>echo "Failed to execute script main"
"Failed to execute script main"

```

**Comment by ahgamut** [comment]
A couple of observations:

1. Freezing numpy seems to cause no issue, but freezing scipy does. I wonder if it has something to do with the DLL locations for each in `site-packages` folder.

2. `from scipy.stats import gamma`, but the warnings show up for unrelated DLLs as well.

**Comment by arcgos** [comment]
Yes almost same issue over here.....
Its working on my friends laptop,  but i tried on a different PC , same issues.. Same steps have been followed in all the scenarios

**Comment by mherrmann** [comment]
Thank you for the detailed example. I tried it on my machine and "unfortunately" everything works flawlessly here.

But I have another idea: Looking at the error you posted more closely, it seems that you are using 64-bit Python 3.6. Does it work when you use [32-bit Python 3.5](https://www.python.org/ftp/python/3.5.4/python-3.5.4.exe) instead?

**Comment by ahgamut** [comment]
No, I am using Python 3.5 32-bit.

```cmd
(mvenv) E:\fbs-tutorial-master\mvenv\Scripts>python
Python 3.5.4 (v3.5.4:3f56838, Aug  8 2017, 02:07:06) [MSC v.1900 32 bit (Intel)]
 on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> exit()
```

Is it possible for me to view verbose output of what `python -m fbs freeze` is doing?

~~**Edit**: This can be due to my version of Windows, maybe? It looks like I am using Windows 7 (64-bit), and using [Dependency Walker](http://www.dependencywalker.com/) on the executable created shows that `ieshims.dll` is not found. Could this be why it crashes on some systems and not others?~~

**Second Edit**: No issue with the above. I downloaded Python3.5 64-bit, followed steps like above, still does not work.

**Comment by ahgamut** [comment]
[PyInstaller seems to have faced (and fixed) a similar issue with PyQt DLLs.](https://github.com/pyinstaller/pyinstaller/issues/2152) 
Maybe a similar issue/workaround is possible here?

**Comment by fredrikaverpil** [comment]
I used to have a lot of missing DLLs but just installing Miniconda3 and adding it to PATH solved all my missing DLL issues. Perhaps worth giving this a try?

Note: I'm using Python 3.6 64-bit (from python.org) for my venv but I also have Conda (3.6 64-bit) on PATH.

**Comment by mherrmann** [comment]
Sorry @ahgamut and @arcgos for the problems. I really want to make this work for you.

That adding `location_of_python_env\Lib\site-packages\scipy\extra-dll\` to the `PATH` environment variable gets rid of the warnings is an interesting/good first step.

You mentioned that you are getting "failed to execute script main". I presume this is a dialog box that opens when you run the application. Does it say anything beyond "failed to execute"?

To get more detailed debug information, you can edit the file [`fbs/freeze/windows.py`](https://github.com/mherrmann/fbs/blob/master/fbs/freeze/windows.py) in the `lib\site-packages` directory of your `venv`: Replace

    def freeze_windows(extra_pyinstaller_args=None):
        ...
        pyinstaller_args = [
            '--windowed',
            '--icon', path('src/main/icons/Icon.ico')
        ]
        ...

by

    def freeze_windows(extra_pyinstaller_args=None):
        ...
        pyinstaller_args = [
            '--debug',
            '--icon', path('src/main/icons/Icon.ico')
        ]
        ...

Then run `python -m fbs freeze` again and run your app from the command line (eg. `target\Tutorial\Tutorial.exe`). This should give more debug output.

**Comment by arcgos** [comment]
@mherrmann  no worries , lets just find out the main issues related to the dlls and resolve them somehow.
pyinstaller --paths C:\Users\zeus\AppData\Local\conda\conda\envs\py35\Lib\site-packages\PyQt5\Qt\bin --add-data 118.h5;. --hidden-import=h5py --hidden-import=h5py.defs --hidden-import=h5py.utils --hidden-import=h5py.h5ac --hidden-import=h5py._proxy project.py
________________________________________________________________________________________
i was having one of the issue due to pyqt5.sip missing dlls so i tried --hidden-import=PyQt5.sip it resolved that particular issue maybe manualy importing all the dll containing modules might work.

**Comment by mherrmann** [comment]
@arcgos it seems @ahgamut was able to resolve the missing DLL issue via [these steps](https://github.com/mherrmann/fbs-tutorial/issues/12#issuecomment-404037411). Second, the commands you just posted seem to be based on Conda. If yes, can we please work together to fix the problems for a vanilla 32 bit Python 3.5 installation and a virtual environment as in the tutorial first please? It rules out errors that may be specific to conda/etc.

**Comment by ahgamut** [comment]
Ah, so adding `--debug` gives this:

```cmd
Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

E:\fbs-tutorial-master\env64\Scripts>activate.bat
(env64) E:\fbs-tutorial-master\env64\Scripts>cd ..\..

(env64) E:\fbs-tutorial-master>python -m fbs freeze
WARNING: The output directory "E:\fbs-tutorial-master\target\Tutorial" and ALL I
TS CONTENTS will be REMOVED! Continue? (y/n)y

(env64) E:\fbs-tutorial-master>cd target\Tutorial

(env64) E:\fbs-tutorial-master\target\Tutorial>Tutorial.exe
[4600] PyInstaller Bootloader 3.x
[4600] LOADER: executable is E:\fbs-tutorial-master\target\Tutorial\Tutorial.exe

[4600] LOADER: homepath is E:\fbs-tutorial-master\target\Tutorial
[4600] LOADER: _MEIPASS2 is NULL
[4600] LOADER: archivename is E:\fbs-tutorial-master\target\Tutorial\Tutorial.ex
e
[4600] LOADER: No need to extract files to run; setting extractionpath to homepa
th
[4600] LOADER: SetDllDirectory(E:\fbs-tutorial-master\target\Tutorial)
[4600] LOADER: Already in the child - running user's code.
[4600] LOADER: Python library: E:\fbs-tutorial-master\target\Tutorial\python35.d
ll
[4600] LOADER: Loaded functions from Python library.
[4600] LOADER: Manipulating environment (sys.path, sys.prefix)
[4600] LOADER: Pre-init sys.path is E:\fbs-tutorial-master\target\Tutorial\base_
library.zip;E:\fbs-tutorial-master\target\Tutorial
[4600] LOADER: sys.prefix is E:\fbs-tutorial-master\target\Tutorial
[4600] LOADER: Setting runtime options
[4600] LOADER: Initializing python
[4600] LOADER: Overriding Python's sys.path
[4600] LOADER: Post-init sys.path is E:\fbs-tutorial-master\target\Tutorial\base
_library.zip;E:\fbs-tutorial-master\target\Tutorial
[4600] LOADER: Setting sys.argv
[4600] LOADER: setting sys._MEIPASS
[4600] LOADER: importing modules from CArchive
[4600] LOADER: extracted struct
[4600] LOADER: callfunction returned...
[4600] LOADER: extracted pyimod01_os_path
[4600] LOADER: callfunction returned...
[4600] LOADER: extracted pyimod02_archive
[4600] LOADER: callfunction returned...
[4600] LOADER: extracted pyimod03_importers
[4600] LOADER: callfunction returned...
[4600] LOADER: Installing PYZ archive with Python modules.
[4600] LOADER: PYZ archive: out00-PYZ.pyz
[4600] LOADER: Running pyiboot01_bootstrap.py
[4600] LOADER: Running pyi_rth_pkgres.py
[4600] LOADER: Running pyi_rth_win32comgenpy.py
[4600] LOADER: Running pyi_rth_multiprocessing.py
[4600] LOADER: Running pyi_rth__tkinter.py
[4600] LOADER: Running pyi_rth_qt5.py
[4600] LOADER: Running pyi_rth_mplconfig.py
[4600] LOADER: Running pyi_rth_mpldata.py
[4600] LOADER: Running main.py
Traceback (most recent call last):
  File "tutorial\main.py", line 1, in <module>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "tutorial\application_context_2.py", line 7, in <module>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "lib\site-packages\scipy\stats\__init__.py", line 345, in <module>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "lib\site-packages\scipy\stats\stats.py", line 171, in <module>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "lib\site-packages\scipy\stats\distributions.py", line 10, in <module>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "lib\site-packages\scipy\stats\_distn_infrastructure.py", line 16, in <mo
dule>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "lib\site-packages\scipy\misc\__init__.py", line 68, in <module>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "lib\site-packages\scipy\interpolate\__init__.py", line 175, in <module>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "lib\site-packages\scipy\interpolate\interpolate.py", line 32, in <module
>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 714, in load_module
    module = loader.load_module(fullname)
  File "interpnd.pyx", line 1, in init scipy.interpolate.interpnd
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "lib\site-packages\scipy\spatial\__init__.py", line 96, in <module>
  File "e:\fbs-tutorial-master\env64\lib\site-packages\PyInstaller\loader\pyimod
03_importers.py", line 714, in load_module
    module = loader.load_module(fullname)
  File "messagestream.pxd", line 5, in init scipy.spatial.qhull
ImportError: No module named 'scipy._lib.messagestream'
[4600] Failed to execute script main
[4600] LOADER: OK.
[4600] LOADER: Cleaning up Python interpreter.

(env64) E:\fbs-tutorial-master\target\Tutorial>
```

6th line from the end: **ImportError: No module named 'scipy._lib.messagestream'**.

[A quick search gives this as an issue with scipy and PyInstaller.](https://github.com/pyinstaller/pyinstaller/issues/2987)

**Comment by ahgamut** [comment]
**Okay, seems like it is fixed.** 

**TL; DR** Add `env_folder\Lib\site-packages\scipy\extra-dll\` to PATH and manually import `scipy._lib.messagestream` wherever you import from scipy.

For the record, `scipy._lib.messagestream` is located in `env_folder\Lib\site-packages\scipy\_lib\messagestream.pyd`

So scipy has:
1. DLLs in a location not usually in PATH / not considered while freezing
2. "hidden imports" (like this `scipy._lib.messagestream`) which earlier versions of PyInstaller don't properly look for.

(@arcgos Maybe tensorflow has similar issues?)

Here are the steps I followed:

(I used Python 3.5.4 64bit, on Windows 7 64bit)

1.  Downloaded a zip file of the repo, followed all the steps in the tutorial and checked that the example program works upon freezing (`success.jpg`)

2. Changed `application_context_2.py` [as per my earlier comment](https://github.com/mherrmann/fbs-tutorial/issues/12#issuecomment-404372123), and tried to freeze again.

(Deleted the `target` folder before every freeze)

3. A whole bunch of DLL warnings show up, but it still freezes. [Changing PyInstaller args to view debug output](https://github.com/mherrmann/fbs-tutorial/issues/12#issuecomment-404757431) as @mherrmann said ~(or by using Dependency Walker)~ shows that the application crashes because it is unable to load the DLLs.

4. I added  `env_folder\Lib\site-packages\scipy\extra-dll\` to PATH and try again. No warnings now (!), but the executable still crashes. Viewing debug output shows that `scipy._lib.messagestream` is missing. (Adding `env_folder\Lib\site-packages\scipy\_lib` to PATH has no effect.)

5. I added `import scipy._lib.messagestream` to `application_context_2.py` and froze it again. There were no warnings, and the executable worked.

I don't think Python (or Windows) being 32bit/64bit is a factor here, but I'll have to check. (My PATH variable is messed up.)

Anyway, following the above steps got the scipy issue solved. Now let me see if I can freeze a larger GUI.

**Comment by arcgos** [comment]
ok

**Comment by mherrmann** [comment]
Thank you very much for sharing your results @ahgamut! I will update fbs and its documentation to make it easier for others to debug such (and similar) issues. And from the PyInstaller issue you referenced, it seems it will be resolved automatically once PyInstaller 3.4 is out.

**Comment by mherrmann** [comment]
There is now a `hidden_imports` setting so you don't have to do `import xyz` in your Python code any more. See the [Manual](https://build-system.fman.io/manual/#dependencies).

**Comment by ScarlettCharlie** [comment]
@arcgos Did you get response because you paid for fbs?

**Comment by mherrmann** [comment]
You're free to harass me @ScarlettCharlie but now beginning to bother fbs users is too much. I'm banning you from this issue tracker. If you create new GitHub accounts to keep harassing me, I will of course block those as well and report you to GitHub for abuse.
